File: content/snippets/go/design_patterns_facade.go
package main

import "fmt"

type Compiler struct{}

func (Compiler) Run() { fmt.Println("compile") }

type TestRunner struct{}

func (TestRunner) Run() { fmt.Println("test") }

type Packager struct{}

func (Packager) Run() { fmt.Println("package") }

type ReleaseFacade struct {
	compiler Compiler
	tests    TestRunner
	packager Packager
}

func (f ReleaseFacade) Deploy() {
	f.compiler.Run()
	f.tests.Run()
	f.packager.Run()
}

func main() {
	ReleaseFacade{}.Deploy()
}

File: content/snippets/typescript/design_patterns_facade.ts
class Compiler {
  run(): void {
    console.log('compile');
  }
}

class TestRunner {
  run(): void {
    console.log('test');
  }
}

class Packager {
  run(): void {
    console.log('package');
  }
}

class ReleaseFacade {
  private readonly compiler = new Compiler();
  private readonly tests = new TestRunner();
  private readonly packager = new Packager();

  deploy(): void {
    this.compiler.run();
    this.tests.run();
    this.packager.run();
  }
}

new ReleaseFacade().deploy();
